storage: TestStoreRangeMergeAbandonedFollowersAutomaticallyGarbageCollected flaky

It failed on https://github.com/cockroachdb/cockroach/pull/31957. The culprit was a timeout, with endless spam:

```
E181031 19:27:12.087176 69259 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot b2ee3b54 at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.136100 69298 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 69451a88 at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.289314 69318 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 3ec1b048 at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.385763 69265 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 647bd927 at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.486789 69249 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 188c170f at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.585974 69234 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 751dc921 at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.635955 69364 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot f8ca11fe at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.786097 69366 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 5498509a at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
E181031 19:27:12.886162 69353 storage/queue.go:803  [raftsnapshot,s1,r2/1:{b-/Max}] snapshot failed: (n3,s3):3: remote couldn't accept Raft snapshot 55fe78af at applied index 13 with error: [n3,s3],r2: cannot apply snapshot: snapshot intersects existing range; initiated GC: [n3,s3,r1/3:/M{in-ax}]
panic: test timed out after 8m0s

goroutine 69379 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/go/src/testing/testing.go:1240 +0xfc
created by time.goFunc
	/usr/local/go/src/time/sleep.go:172 +0x44
```

--- a/pkg/storage/client_merge_test.go
+++ b/pkg/storage/client_merge_test.go
@@ -2022,8 +2022,6 @@ func TestStoreRangeMergeAbandonedFollowers(t *testing.T) {
 func TestStoreRangeMergeAbandonedFollowersAutomaticallyGarbageCollected(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 
-	t.Skip("https://github.com/cockroachdb/cockroach/issues/32062")
-
 	ctx := context.Background()
 	storeCfg := storage.TestStoreConfig(nil)
 	storeCfg.TestingKnobs.DisableReplicateQueue = true
@@ -2038,8 +2036,19 @@ func TestStoreRangeMergeAbandonedFollowersAutomaticallyGarbageCollected(t *testi
 		t.Fatal(err)
 	}
 
-	// Make store2 the leaseholder for the RHS.
+	// Make store2 the leaseholder for the RHS and wait for the lease transfer to
+	// apply.
 	mtc.transferLease(ctx, rhsDesc.RangeID, 0, 2)
+	testutils.SucceedsSoon(t, func() error {
+		rhsRepl, err := store2.GetReplica(rhsDesc.RangeID)
+		if err != nil {
+			return err
+		}
+		if !rhsRepl.OwnsValidLease(mtc.clock.Now()) {
+			return errors.New("store2 does not own valid lease for rhs range")
+		}
+		return nil
+	})
 
 	// Start dropping all Raft traffic to the LHS replica on store2 so that it
 	// won't be aware that there is a merge in progress.

